File: oxidize-golang/core/autotune/apply.go
package autotune

import "github.com/Zapdev-labs/oxidize/golang/core/kv_cache"

// PlanOverrides holds per-flag autotune recommendations for CLI/server apply.
type PlanOverrides struct {
	Threads        *int
	CtxSize        *int
	NGPULayers     *int
	LayerCache     *int
	LayerWise      *bool
	Mmap           *bool
	Mlock          *bool
	MmapHugepages  *bool
	MmapPrefetch   *bool
	RAMOffload     *bool
	CPUOptimized   *bool
	TurboQuant     *bool
	Pipeline       *string
	DecodeTile     *int
}

// OverridesFromPlan converts a tuning plan into flag overrides.
func OverridesFromPlan(plan *TuningPlan) PlanOverrides {
	pipeline := pipelineString(plan.Pipeline)
	turbo := plan.KVQuantization == kv_cache.QuantTurboQuant
	cpuOpt := false
	decodeTile := (*int)(nil)
	if plan.DecodeTileTokens > 0 {
		dt := plan.DecodeTileTokens
		decodeTile = &dt
	}
	return PlanOverrides{
		Threads:       &plan.Threads,
		CtxSize:       &plan.CtxSize,
		NGPULayers:    &plan.NGPULayers,
		LayerCache:    &plan.LayerCache,
		LayerWise:     &plan.LayerWise,
		Mmap:          &plan.Mmap,
		Mlock:         &plan.Mlock,
		MmapHugepages: &plan.MmapHugepages,
		MmapPrefetch:  &plan.MmapPrefetch,
		RAMOffload:    &plan.Mlock,
		CPUOptimized:  &cpuOpt,
		TurboQuant:    &turbo,
		Pipeline:      &pipeline,
		DecodeTile:    decodeTile,
	}
}

func pipelineString(mode PipelineMode) string {
	switch mode {
	case PipelineSequential:
		return "sequential"
	case PipelineContinuous:
		return "continuous"
	case PipelinePaged:
		return "paged"
	case PipelineAsymmetric:
		return "asymmetric"
	default:
		return "sequential"
	}
}

File: oxidize-python/oxidize_python/core/autotune/apply.py
"""Apply autotune plans to CLI options."""

from __future__ import annotations

from dataclasses import dataclass

from oxidize_python.core.autotune.rules import PipelineMode, TuningPlan
from oxidize_python.core.kv_cache import Quantization as KvQuant


@dataclass
class PlanOverrides:
    threads: int | None = None
    ctx_size: int | None = None
    n_gpu_layers: int | None = None
    layer_cache: int | None = None
    layer_wise: bool | None = None
    mmap: bool | None = None
    paged: bool | None = None
    turboquant: bool | None = None
    pipeline: str | None = None


def overrides_from_plan(plan: TuningPlan) -> PlanOverrides:
    pipeline = {
        PipelineMode.SEQUENTIAL: "sequential",
        PipelineMode.CONTINUOUS: "continuous",
        PipelineMode.PAGED: "paged",
        PipelineMode.ASYMMETRIC: "asymmetric",
    }[plan.pipeline]
    return PlanOverrides(
        threads=plan.threads,
        ctx_size=plan.ctx_size,
        n_gpu_layers=plan.n_gpu_layers,
        layer_cache=plan.layer_cache,
        layer_wise=plan.layer_wise,
        mmap=plan.mmap,
        paged=plan.pipeline == PipelineMode.PAGED,
        turboquant=plan.kv_quantization == KvQuant.TURBOQUANT,
        pipeline=pipeline,
    )
